### DISCUSSION TITLE
Accessing the static members on a bounded type parameter

### DISCUSSION THREAD

**Discussion by bwburnsides** [microsoft/pyright#7814] [Q&A]
So I've run into a problem which I suspect may be a brick wall in the capabilities of the Python type annotation system. Given a type parameter `T`, bounded to a protocol `P` that affords a `V: ClassVar[...]`, how do I access `V`? Here's a more concrete example:
```py
class ExprPhase(t.Protocol):
    XLiteral: t.ClassVar[t.TypeAliasType]
    XVariable: t.ClassVar[t.TypeAliasType]
    XOther: t.ClassVar[T.TypeAliasType]

class XLiteral[T: ExprPhase]:
    type Phase = T.XLiteral # <--- 'Cannot access member "XLiteral" for type "TypeVar". Member "XLiteral" is unknown.'

class XVariable[T: ExprPhase]:
    type Phase = T.XVariable  # ERROR: See above.

class XOther[T: ExprPhase]:
    type Phase = T.XOther  # ERROR: See above.
```
I understand why this error is occurring, but I'm not sure how to get around it...

I recognize that I may be trying to fix a square peg into a round whole. I'm experimenting with translating the "Trees that Grow" design pattern for compiler tree decoration from Haskell to Python, using the [comment found here for reference](https://www.reddit.com/r/ProgrammingLanguages/comments/dsbduz/comment/f6ok66c/). **More generally, I am trying to represent Haskell type families in Python.**

**Comment by erictraut** [comment]
I don't understand what you're trying to do here, so I don't have any suggestions.

You asked "Given a type parameter `T`, bounded to a protocol `P`"... There is no such thing as "binding a type parameter to a protocol" in Python, so I'm confused by that statement. Is your intent to add an upper-bound constraint for the type parameter?

It appears you're trying to access attributes of type variables which isn't allowed in the Python type system. I'm not even sure what that would mean.

You're also using `TypeAliasType` in a manner that is very odd. This type is a runtime type used to implement type aliases. It generally shouldn't appear in type annotations.

If you have questions about the Python static type system in general, the [Python typing forum](https://discuss.python.org/c/typing/32) is a good place to post questions.

**Reply by bwburnsides** [reply]
Yes, `ExprPhase` is intended to be an upper bound on `T`. I accidentally omitted that in my sample, sorry about that... its added now.

I'm not sure that `TypeAliasType` is the correct type to use in this case, but for context, a type which implements this protocol would ideally look something like this. Perhaps the correct type would be `t.ClassVar[t.TypeAlias]`.
```py
class Example:
    type XLiteral = int
    type XVariable = int
    type XOther = int
```

**Reply by bwburnsides** [reply]
With the upper bound in place, my intent is that the upper bound providing the type aliases `XLiteral`, etc would allow me to access `T.XLiteral`. However since `.` is a runtime operation, `T` is being treated as a `TypeVar` (correctly of course). I'm looking for a way to say "the type found in the namespace of whatever type that's provided as a generic argument".

**Reply by erictraut** [reply]
I'm still not clear on what you're trying to do. It sounds like maybe you're trying to use a form of higher-kinded types (HKTs)? That concept doesn't exist in the Python type system.

You may find it useful to review the Python typing reference and specification [here](https://typing.readthedocs.io/en/latest/) to get a better understanding of the facilities that are present in the Python type system.

**Reply by bwburnsides** [reply]
Yes, this is in the neighborhood of HKTs... I'm pretty familiar with the typing facilities but I was hoping for something clever 🤠. I think simply put, I am trying to do the equivalent of this. I know my terminology above may not be correct still.
```cpp
template <typename T>
struct XOther {
    typedef typename T::XOther Phase;
};
```
